### PR TITLE
Buildclusters queue

### PR DESCRIPTION
**Pull request by parthasl**
## Context

Build scalable Screwdriver build infrastructure.

## Objective

Design proposal for multiple build clusters.

1. Running builds in multiple build clusters to support high availability
2. Users can bring in their own build clusters for their own specific infrastructure needs.
3. Support build cluster in AWS

## References

#1319

### CODE REVIEW

**Comment by tkyi** [comment]
Why is this a separate PR? All of the discussion is in https://github.com/screwdriver-cd/screwdriver/pull/1324

File: design/build-clusters.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+## Flow
+### SD Validator
+
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### Screwdriver API to Scheduler service
+
+	1. Screwdriver UI / PR commit / Merge triggers new build via Screwdriver API
+	2. Screwdriver API inturn calls Scheduler service with appropriate build details to schedule a job
+	3. Authentication via JWT token (refer Authorization section)
```
Comment (minz1027) [inline design/build-clusters.md]: Based on the discuss we have, it's more like, API still enqueues to redis directly as what we have today. It's just that the current queue-worker moves to the services cluster and became a scheduler. So API doesn't really need to talk to scheduler. Scheduler will pick up jobs from redis and put jobs to proper queue.

File: design/build-clusters.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	4. Scheduler service queries `buildClusters` cache for active records with cluster name from build info and validate if job can be scheduled in appropriate buildCluster queue
+	5. one (or) more record exist, then assign job to the queue identified by generating a random number within given boundaries, which is the returned list size of records
+	6. no records, then query `clusters` table for active records with managedBy=screwdriver
+	7. repeat step #4
+	8. Update build info with cluster and queue details
+
+### Queue worker to Scheduler service
```
Comment (minz1027) [inline design/build-clusters.md]: This is more like Queue worker to Queue, not scheduler, it doesn't need to talk to scheduler

File: design/build-clusters.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+3. SD admin to create queue based on #1
+4. SD admin authorize build cluster user and queue 
+
+
+## Flow
+### SD Validator
+
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### Screwdriver API to Scheduler service
```
Comment (minz1027) [inline design/build-clusters.md]: This will be API -> Redis and Scheduler -> Redis

File: design/build-clusters.md (hunk 1 above)
Comment (parthasl) [reply design/build-clusters.md]: @minz1027 are we not planning to do the scheduler service refactoring as well ? combining the scheduler and executor queue ?

File: design/build-clusters.md (hunk 2 above)
Comment (parthasl) [reply design/build-clusters.md]: @minz1027 ack, will change

**Comment by parthasl** [comment]
> Why is this a separate PR? All of the discussion is in #1324

this is a different design approach from earlier one.

File: design/build-clusters.md (hunk 3 above)
Comment (parthasl) [reply design/build-clusters.md]: will wait for decision on https://github.com/screwdriver-cd/screwdriver/pull/1357#discussion_r226763615

File: design/build-clusters.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### API to Redis
+
+	1. No change in existing implementation.
+
+### Master queue worker to Queue 
+
+	1. Master queue worker queries `buildClusters` cache for active records with cluster name from build info 
+	2. Validates if build job can be scheduled in specified buildCluster queue by validating scmContext + scmOrganization access and validate if blockedBy and other checks are success.
```
Comment (minz1027) [inline design/build-clusters.md]: We shouldn't need to do the check any more, it should either be done by the validator or model when we tries to create the build. When it comes here, we can assume it has proper permission

File: design/build-clusters.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+| 4 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 0 | external | identity@oath.com |
+| 5 | iOS | github:git.ouroath.com | iOS_org1, iOS_org2 | 1 | external | ios@oath.com |
+
+### Cache server to store active *buildClusters* in memory when the service boot up. 
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url | Description
+| --- | --- | ---
+| `POST` | ` /buildClusters ` | ` { "name":"iOS", "scmContext":"github:git.ouroath.com", "scmOrganizations": "iOS_org1", "isActive":1, "managedBy": "screwdriver" } `
```
Comment (minz1027) [inline design/build-clusters.md]: scmOrganizations should be an array

File: design/build-clusters.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | yes | no | no | |
+| `scmContext` | text(200) | no | yes | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
```
Comment (minz1027) [inline design/build-clusters.md]: can we just use true and false for the value?

File: design/build-clusters.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### API to Redis
+
+	1. No change in existing implementation.
+
+### Model
+
+	1. Query `buildClusters` cache for active records with cluster name from build info
```
Comment (minz1027) [inline design/build-clusters.md]: Query `buildClusters` table.

File: design/build-clusters.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
+| `managedByEmail` | text(100) | yes | no | no | cluster admin email for communications |
+
+Unique constraint: `name + isActive` 
+
+#### Sample record
+
+| id | name | scmContext | scmOrganizations | isActive | managedBy | managedByEmail 
+| --- | --- | --- | --- | --- | --- | --- | 
+| 1 | gq1 | github:git.ouroath.com | null | true | screwdriver | sd@oath.com |
```
Comment (minz1027) [inline design/build-clusters.md]: @jithin1987 can we specify any internal stuff here?

Comment (jithin1987) [reply design/build-clusters.md]: please update this to public one

File: design/build-clusters.md (hunk 7 above)
Comment (parthasl) [reply design/build-clusters.md]: fixed

File: design/build-clusters.md (hunk 8 above)
Comment (parthasl) [reply design/build-clusters.md]: fixed

File: design/build-clusters.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+	
+### API to Redis
+
+	1. No change in existing implementation.
+
+### Model
+
+	1. Query `buildClusters` table for active records with cluster name from build info 
+	2. Validates if build job can be scheduled in specified buildCluster queue by validating scmContext + scmOrganization access. 
+	3. one (or) more record exist, then assign job to the queue identified by generating a random number within given boundaries, which is the returned list size of records
```
Comment (jithin1987) [inline design/build-clusters.md]: Also we need to capture that this works as long as build nodes in cluster is homogeneous. We should also look to add weighs to individual clusters if there is significant difference in capacity between them.

Comment (parthasl) [reply design/build-clusters.md]: consumer fetch count per cluster can be configured individually based on their capacity. multiple clusters can listen to one queue and pull messages based on the configured fetch count.

Comment (parthasl) [reply design/build-clusters.md]: as discussed, included weighted percentage, but percentage should be adjusted automatically (or) manually when a cluster is made inactive.

File: design/build-clusters.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | yes | no | no | |
+| `scmContext` | text(200) | no | yes | no | |
+| `isActive` | boolean | no | no | no | *false or true* |
+| `defaultCluster` | boolean | no | no | no | *true - cluster maintained by screwdriver team, false - cluster maintained by  external team * |
```
Comment (tkyi) [inline design/build-clusters.md]: should this be `managedByScrewdriver` to match the data-schema?
though I prefer `defaultCluster`

File: design/build-clusters.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+| --- | --- | --- | --- | --- | --- | --- | ---
+| 1 | gq1 | github:github.com | null | true | true | sd@foo.com | 70
+| 2 | bf1 | github:github.com | null | false | true | sd@foo.com | 30
+| 3 | iOS | github:github.com | [iOS_org1, iOS_org2] | true | false | ios@foo.com | 100
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url | Description
+| --- | --- | ---
+| `POST` | ` /buildClusters ` | ` { "name":"iOS", "scmContext":"github:github.com", "scmOrganizations": "[iOS_org1, iOS_org2]", "isActive":true, "defaultCluster": true } `
```
Comment (tkyi) [inline design/build-clusters.md]: should be `/buildclusters`, lowercase `c`

File: design/build-clusters.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+	
+
+As part of cluster onboarding process for above both options cluster admin should register their build cluster information with Screwdriver. Below details are required to register a build cluster.
+
+	1. Cluster name - Cluster name provided by the client
+	2. isActive - Cluster status whether its active or inactive. This will be used to route / pause. Initially this will be a manual update of cluster health.
+	3. Authorization - Authorize using user credentials and queue details.
+	4. SCM Context - git url (github.com). Applicable to only explicit build clusters which are non SD maintained.
+	5. SCM Organization - git organizations. Will be used to validate if job has permission to run on build cluster which is requesting. Applicable to only explicit build clusters which are non SD maintained.
+	6. managedByScrewdriver (true / false) - if true Cluster is maintained by screwdriver team, false cluster is maintained by external team.
```
Comment (Filbird) [inline design/build-clusters.md]: lowercase `Cluster` -> `cluster`. 

Can change `screwdriver` -> `SD` for consistency.

File: design/build-clusters.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
```
Comment (Filbird) [inline design/build-clusters.md]: YAML indentation is a bit off here.

File: design/build-clusters.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
```
Comment (Filbird) [inline design/build-clusters.md]: YAML indentation for `addresses` and `statuses`

File: design/build-clusters.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | yes | no | no | |
+| `scmContext` | text(200) | no | yes | no | |
+| `isActive` | boolean | no | no | no | *false or true* |
+| `managedByScrewdriver` | boolean | no | no | no | *true - cluster maintained by screwdriver team, false - cluster maintained by  external team * |
```
Comment (Filbird) [inline design/build-clusters.md]: Markdown for `*true - cluster maintained by screwdriver team, false - cluster maintained by  external team *` is not being evaluated properly.

**Comment by vengopal** [comment]
👍

**Pull request merged**